lib/builder/builder.go

```diff
@@ -296,49 +335,82 @@ func (b *Builder) initServices() (err error) {
+	// a special case only for building from local packages
+	if b.StateDir != "" {
+		return localenv.NewLocalEnvironment(localenv.LocalEnvironmentArgs{
+			StateDir:         b.StateDir,
+			LocalKeyStoreDir: b.StateDir,
+			Insecure:         b.Insecure,
+		})
+	}
+	// otherwise use default locations for cache / key store
+	cacheDir, err := ensureCacheDir(b.Repository)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return localenv.NewLocalEnvironment(localenv.LocalEnvironmentArgs{
+		StateDir: cacheDir,
+		Insecure: b.Insecure,
+	})
+}
+
+// getSyncer returns a new syncer instance for this builder
+func (b *Builder) getSyncer() (Syncer, error) {
+	if b.StateDir != "" {
+		apps, err := b.Env.AppServiceLocal(localenv.AppConfig{})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return NewPackSyncer(b.Env.Packages, apps, b.StateDir), nil
+	}
+	return b.NewSyncer(b)
+}
+
 // checkVersion makes sure that the tele version is compatible with the selected
 // runtime version
 //
-// Compatibility is defined as "tele version must be >= runtime version"
-func (b *Builder) checkVersion() error {
+// Versions are compatible when they have equal major and minor components.
+func (b *Builder) checkVersion(runtimeVersion *semver.Version) error {
 	teleVersion, err := semver.NewVersion(version.Get().Version)
 	if err != nil {
 		return trace.Wrap(err, "failed to determine tele version")
 	}
-	runtimeLoc := b.Manifest.Base()
-	if runtimeLoc == nil {
-		return trace.BadParameter("failed to determine runtime version, make "+
-			"sure your application type is %q", schema.KindBundle)
-	}
-	runtimePackage, err := b.Packages.ReadPackageEnvelope(*runtimeLoc)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	runtimeVersion, err := semver.NewVersion(runtimePackage.Locator.Version)
-	if err != nil {
-		return trace.Wrap(err, "invalid runtime version: %v",
-			runtimePackage.Locator.Version)
-	}
-	if teleVersion.LessThan(*runtimeVersion) {
-		// truncate meta information
-		withoutMeta := fmt.Sprintf("%v.%v.%v",
-			runtimeVersion.Major, runtimeVersion.Minor, runtimeVersion.Patch)
+	if teleVersion.Major != runtimeVersion.Major || teleVersion.Minor != runtimeVersion.Minor {
 		return trace.BadParameter(
 			`The version of the tele binary (%v) is not compatible with the selected runtime (%v).
 
-Please upgrade your Gravity tools to match the runtime version (curl https://get.gravitational.io/telekube/install/%v | bash), or specify an appropriate runtime version in your application manifest. You can view available runtimes using "tele ls --runtimes" and pick the one that matches your tele version.`, teleVersion, runtimeVersion, withoutMeta)
+To remediate the issue you can do one of the following:
+
+ * Use a different version of tele that is compatible with your selected runtime (latest %v.%v.x version is recommended).
+ * Pin runtime version in application manifest to the version compatible with your tele version (latest %v.%v.x version is recommended).
```

Comment:
Also, is it clear to a user what exactly `pinning` and especially `unpinning` a runtime version in the manifest means? I would say either:
  * `specify the desired version of runtime explicitly` (maybe even with a reference to documentation) 
  * or `do not specify the runtime version explicitly and let tele determine the version automatically`